encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/TypeToken.java

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
+
+public abstract class TypeToken<T> {
+
+
+  @NonNull
+  public static <T> TypeToken<T> of(@NonNull Safe<T> token) {
+    Type type = token.getType();
+    return of(type);
+  }
+
+  @NonNull
+  private static <T> TypeToken<T> of(@NonNull Type type) {
+    if(type instanceof WildcardType) {
+      return of(((WildcardType) type).getUpperBounds()[0]);
```

Comment:
we have to handle the case when `getUpperBounds()` is empty and throw an exception. This could happen if I specify `TypeToken.of(new Safe<List<? super Foo>>() {})`

Reply:
It turns out that `upperBounds()` of `TypeToken.of(new Safe<List<? super Foo>>() {})` is `Object.class`.  And we cannot compare it with `Object.class`, since the `upperBound` of `TypeToken.of(new Safe<List<? extend Object>>() {})` is `Object.class` as well.
Instead, `lowerBouds()` is empty when I specify `TypeToken.of(new Safe<List<? extend Foo>>() {})`.
Since bounded wild card type only has those two options, we can prevent it as follow:
```
      if(((WildcardType) type).getLowerBounds().length == 0) {
        return of(((WildcardType) type).getUpperBounds()[0]);
      }
      throw new RuntimeException("<? super T> is not supported");
```
@vkryachko 
Wdyt?